File: apps/go/simulator/universe.go
package simulator

import (
	"fmt"
	"math/rand"
)

// EpochName maps epoch tick thresholds to human-readable names.
type EpochInfo struct {
	Name        string
	StartTick   int
	Description string
	Symbol      string
}

// Epochs defines the 13 cosmic epochs in chronological order.
var Epochs = []EpochInfo{
	{"Planck", PlanckEpoch, "Quantum gravity dominates; spacetime foam", "*"},
	{"Inflation", InflationEpoch, "Exponential expansion of spacetime", "~"},
	{"Electroweak", ElectroweakEpoch, "Electroweak symmetry breaking; W/Z bosons acquire mass", "!"},
	{"QuarkGluon", QuarkEpoch, "Quark-gluon plasma; free quarks and gluons", "#"},
	{"Hadron", HadronEpoch, "Quarks confine into protons and neutrons", "@"},
	{"Nucleosynthesis", NucleosynthesisEpoch, "Light nuclei form: H, He, Li", "%"},
	{"Recombination", RecombinationEpoch, "Electrons captured by nuclei; atoms form", "&"},
	{"StarFormation", StarFormationEpoch, "First stars ignite; heavy element forging", "+"},
	{"SolarSystem", SolarSystemEpoch, "Protoplanetary disk collapses; planets form", "O"},
	{"EarthFormation", EarthEpoch, "Rocky planet cools; oceans condense", "o"},
	{"LifeEmergence", LifeEpoch, "First protocells arise from prebiotic chemistry", "."},
	{"DNAEvolution", DNAEpoch, "DNA replication, mutation, and natural selection", "="},
	{"Present", PresentEpoch, "Complex life; billions of years of evolution", "^"},
}

// EpochResult stores the outcome of simulating one epoch.
type EpochResult struct {
	EpochName    string
	StartTick    int
	EndTick      int
	Temperature  float64
	ParticleCount int
	AtomCount    int
	MoleculeCount int
	CellCount    int
	Events       []string
}

// Universe is the top-level orchestrator that drives the entire simulation
// from the Planck epoch through to the present.
type Universe struct {
	Field     *QuantumField
	Atoms     *AtomicSystem
	Chemistry *ChemicalSystem
	Biosphere *Biosphere
	Env       *Environment

	Age           int // current tick
	ScaleFactor   float64
	Results       []EpochResult
	rng           *rand.Rand

	// Callback invoked after each epoch completes. May be nil.
	OnEpochComplete func(result EpochResult)

	// Callback invoked each simulation tick. May be nil.
	// Receives the current epoch name, epoch index, and tick number.
	OnTick func(epochName string, epochIndex int, tick int)
}

// NewUniverse creates a brand new universe ready for simulation.
func NewUniverse(seed int64) *Universe {
	rng := rand.New(rand.NewSource(seed))
	field := NewQuantumField(TPlanck, rng)
	atoms := NewAtomicSystem(rng)
	chem := NewChemicalSystem(atoms, rng)
	env := NewEnvironment(TPlanck, rng)

	return &Universe{
		Field:       field,
		Atoms:       atoms,
		Chemistry:   chem,
		Env:         env,
		ScaleFactor: 1.0,
		Results:     make([]EpochResult, 0, len(Epochs)),
		rng:         rng,
	}
}

// Run executes the full simulation across all 13 epochs.
func (u *Universe) Run() {
	for i, epoch := range Epochs {
		endTick := PresentEpoch
		if i+1 < len(Epochs) {
			endTick = Epochs[i+1].StartTick
		}
		result := u.simulateEpoch(epoch, endTick, i)
		u.Results = append(u.Results, result)
		if u.OnEpochComplete != nil {
			u.OnEpochComplete(result)
		}
	}
}

// simulateEpoch runs the simulation from the current age to endTick.
func (u *Universe) simulateEpoch(epoch EpochInfo, endTick int, epochIndex int) EpochResult {
	result := EpochResult{
		EpochName: epoch.Name,
		StartTick: epoch.StartTick,
		EndTick:   endTick,
		Events:    make([]string, 0),
	}

	// Determine step size to keep each epoch tractable.
	totalTicks := endTick - epoch.StartTick
	step := 1
	if totalTicks > 500 {
		step = totalTicks / 200
	}
	if step < 1 {
		step = 1
	}

	for tick := epoch.StartTick; tick < endTick; tick += step {
		u.Age = tick
		u.Env.Update(tick)
		u.Field.Temperature = u.Env.Temperature

		switch {
		// --- Planck & Inflation: vacuum fluctuations produce particles ---
		case tick < ElectroweakEpoch:
			for i := 0; i < 5; i++ {
				u.Field.VacuumFluctuation()
			}
			u.Field.Evolve(0.01)
			u.ScaleFactor *= 1.001

		// --- Electroweak: symmetry breaking, more pair production ---
		case tick < QuarkEpoch:
			for i := 0; i < 3; i++ {
				u.Field.VacuumFluctuation()
			}
			// Extra pair production from thermal energy
			energy := u.Env.Temperature * KB * 10
			u.Field.PairProduction(energy)
			u.Field.Evolve(0.01)
			u.Field.Cool(0.999)

		// --- Quark-Gluon Plasma: quarks and gluons free ---
		case tick < HadronEpoch:
			u.Field.VacuumFluctuation()
			u.Field.Evolve(0.01)
			u.Field.Cool(0.999)

		// --- Hadron Epoch: quark confinement ---
		case tick < NucleosynthesisEpoch:
			u.Field.Cool(0.998)
			hadrons := u.Field.QuarkConfinement()
			if len(hadrons) > 0 {
				result.Events = append(result.Events,
					fmt.Sprintf("Quark confinement: %d hadrons formed", len(hadrons)))
			}
			u.Field.Evolve(0.01)

		// --- Big Bang Nucleosynthesis: light nuclei ---
		case tick < RecombinationEpoch:
			u.Field.Cool(0.9995)
			// Count protons & neutrons for nucleosynthesis
			protons, neutrons := 0, 0
			for _, p := range u.Field.Particles {
				if p.Type == ParticleProton {
					protons++
				}
				if p.Type == ParticleNeutron {
					neutrons++
				}
			}
			if protons > 1 && neutrons > 1 {
				newAtoms := u.Atoms.Nucleosynthesis(protons/4, neutrons/4)
				if len(newAtoms) > 0 {
					result.Events = append(result.Events,
						fmt.Sprintf("Nucleosynthesis: %d nuclei formed", len(newAtoms)))
				}
			}

		// --- Recombination: atoms form ---
		case tick < StarFormationEpoch:
			u.Atoms.Temperature = u.Env.Temperature
			newAtoms := u.Atoms.Recombination(u.Field)
			if len(newAtoms) > 0 {
				result.Events = append(result.Events,
					fmt.Sprintf("Recombination: %d atoms formed", len(newAtoms)))
			}

		// --- Star Formation: heavy elements ---
		case tick < SolarSystemEpoch:
			newElements := u.Atoms.StellarNucleosynthesis(TStellarCore)
			if len(newElements) > 0 {
				result.Events = append(result.Events,
					fmt.Sprintf("Stellar nucleosynthesis: %d heavy elements forged", len(newElements)))
			}

		// --- Solar System Formation ---
		case tick < EarthEpoch:
			// Produce some extra atoms for planet building
			if len(u.Atoms.Atoms) < 80 {
				for _, z := range []int{1, 1, 1, 1, 6, 7, 8, 8, 14, 26} {
					u.Atoms.Atoms = append(u.Atoms.Atoms, NewAtom(z, 0))
				}
				result.Events = append(result.Events, "Accretion disk: elements aggregating")
			}

		// --- Earth Formation: chemistry begins ---
		case tick < LifeEpoch:
			// Ensure we have enough CHON atoms for prebiotic chemistry
			if len(u.Atoms.Atoms) < 120 {
				for _, z := range []int{1, 1, 1, 1, 1, 1, 6, 6, 6, 7, 7, 8, 8, 8} {
					u.Atoms.Atoms = append(u.Atoms.Atoms, NewAtom(z, 0))
				}
			}
			waters := u.Chemistry.FormWater()
			if len(waters) > 0 {
				result.Events = append(result.Events,
					fmt.Sprintf("Oceans forming: %d water molecules", len(waters)))
			}
			u.Chemistry.FormMethane()
			u.Chemistry.FormAmmonia()
			u.Chemistry.CatalyzedReaction(u.Env.Temperature, false)

		// --- Life Emergence: first cells ---
		case tick < DNAEpoch:
			// Keep enriching with atoms for amino acids / nucleotides
			if u.rng.Float64() < 0.3 {
				for _, z := range []int{1, 1, 6, 7, 8} {
					u.Atoms.Atoms = append(u.Atoms.Atoms, NewAtom(z, 0))
				}
			}
			u.Chemistry.CatalyzedReaction(u.Env.Temperature, true)

			// Bootstrap biosphere when enough nucleotides exist
			if u.Biosphere == nil && u.Chemistry.NucleotideCount >= 2 {
				u.Biosphere = NewBiosphere(5, 60, u.rng)
				result.Events = append(result.Events,
					"LIFE EMERGES: first protocells self-assemble!")
			}
			if u.Biosphere != nil {
				u.Biosphere.Step(
					u.Env.ThermalEnergy(),
					u.Env.UVIntensity,
					u.Env.CosmicRayFlux,
					u.Env.Temperature,
				)
			}

		// --- DNA Evolution & Present ---
		default:
			if u.Biosphere == nil {
				u.Biosphere = NewBiosphere(5, 60, u.rng)
				result.Events = append(result.Events,
					"LIFE EMERGES: first protocells self-assemble!")
			}
			u.Biosphere.Step(
				u.Env.ThermalEnergy(),
				u.Env.UVIntensity,
				u.Env.CosmicRayFlux,
				u.Env.Temperature,
			)
		}

		if u.OnTick != nil {
			u.OnTick(epoch.Name, epochIndex, tick)
		}
	}

	// Snapshot state at end of epoch
	result.Temperature = u.Env.Temperature
	result.ParticleCount = len(u.Field.Particles)
	result.AtomCount = len(u.Atoms.Atoms)
	result.MoleculeCount = len(u.Chemistry.Molecules)
	if u.Biosphere != nil {
		result.CellCount = len(u.Biosphere.Cells)
	}

	return result
}

// Snapshot captures the current state of the universe for external consumers.
type Snapshot struct {
	Epoch         string       `json:"epoch"`
	EpochIndex    int          `json:"epoch_index"`
	Tick          int          `json:"tick"`
	Temperature   float64      `json:"temperature"`
	Particles     int          `json:"particles"`
	Atoms         int          `json:"atoms"`
	Molecules     int          `json:"molecules"`
	Cells         int          `json:"cells"`
	TotalEpochs   int          `json:"total_epochs"`
	ParticlePos   [][3]float64 `json:"particle_pos,omitempty"`
	ParticleTypes []string     `json:"particle_types,omitempty"`
}

// TakeSnapshot returns a Snapshot of the current universe state.
// If includePositions is true, particle positions and types are included.
func (u *Universe) TakeSnapshot(epochName string, epochIndex int, includePositions bool) Snapshot {
	s := Snapshot{
		Epoch:       epochName,
		EpochIndex:  epochIndex,
		Tick:        u.Age,
		Temperature: u.Env.Temperature,
		Particles:   len(u.Field.Particles),
		Atoms:       len(u.Atoms.Atoms),
		Molecules:   len(u.Chemistry.Molecules),
		TotalEpochs: len(Epochs),
	}
	if u.Biosphere != nil {
		s.Cells = len(u.Biosphere.Cells)
	}
	if includePositions {
		maxParticles := 200
		count := len(u.Field.Particles)
		if count > maxParticles {
			count = maxParticles
		}
		s.ParticlePos = make([][3]float64, count)
		s.ParticleTypes = make([]string, count)
		for i := 0; i < count; i++ {
			s.ParticlePos[i] = u.Field.Particles[i].Position
			s.ParticleTypes[i] = u.Field.Particles[i].Type.String()
		}
	}
	return s
}

// Summary returns a human-readable summary of the simulation.
func (u *Universe) Summary() string {
	s := ""
	s += fmt.Sprintf("Universe age: %d ticks\n", u.Age)
	s += fmt.Sprintf("Particles: %d (created: %d, annihilated: %d)\n",
		len(u.Field.Particles), u.Field.TotalCreated, u.Field.TotalAnnihilated)
	s += fmt.Sprintf("Total field energy: %.2f\n", u.Field.TotalEnergy())
	s += fmt.Sprintf("Atoms: %d\n", len(u.Atoms.Atoms))
	s += fmt.Sprintf("Element census: %v\n", u.Atoms.ElementCounts())
	s += fmt.Sprintf("Molecules: %d (water: %d, amino acids: %d, nucleotides: %d)\n",
		len(u.Chemistry.Molecules), u.Chemistry.WaterCount, u.Chemistry.AminoAcidCount, u.Chemistry.NucleotideCount)
	s += fmt.Sprintf("Chemical reactions: %d\n", u.Chemistry.ReactionsOccurred)
	if u.Biosphere != nil {
		s += fmt.Sprintf("Living cells: %d (born: %d, died: %d)\n",
			len(u.Biosphere.Cells), u.Biosphere.TotalBorn, u.Biosphere.TotalDied)
		s += fmt.Sprintf("Generations: %d\n", u.Biosphere.Generation)
		s += fmt.Sprintf("Average fitness: %.4f\n", u.Biosphere.AverageFitness())
		s += fmt.Sprintf("Average GC content: %.4f\n", u.Biosphere.AverageGCContent())
		s += fmt.Sprintf("Total mutations: %d\n", u.Biosphere.TotalMutations())
	}
	s += fmt.Sprintf("Environment: T=%.2f K, habitable=%v\n", u.Env.Temperature, u.Env.IsHabitable())
	s += fmt.Sprintf("Environmental events witnessed: %d\n", len(u.Env.EventHistory))
	return s
}

File: apps/cpp/simulator/universe.cpp
#include "universe.hpp"

#include <cmath>
#include <sstream>
#include <iomanip>
#include <memory>

namespace sim {

Universe::Universe()
    : quantumField(T_PLANCK),
      atomicSystem(T_RECOMBINATION),
      chemicalSystem(atomicSystem),
      environment(T_PLANCK) {}

EpochState Universe::snapshot(const std::string& name,
                              const std::string& details) const {
    EpochState state;
    state.epochName     = name;
    state.tick          = tick_;
    state.temperature   = environment.currentTemperature();
    state.totalEnergy   = quantumField.totalEnergy();
    state.particleCount = static_cast<int>(quantumField.particles.size());
    state.atomCount     = static_cast<int>(atomicSystem.atoms.size());
    state.moleculeCount = static_cast<int>(chemicalSystem.molecules.size());
    state.cellCount     = biosphere ? static_cast<int>(biosphere->cells.size()) : 0;
    state.scaleFactor   = environment.scaleFactor();
    state.details       = details;
    return state;
}

void Universe::simulate(std::function<void(const EpochState&)> onEpoch) {
    history_.clear();

    using EpochFn = EpochState (Universe::*)();
    EpochFn epochFns[] = {
        &Universe::simulatePlanck,
        &Universe::simulateInflation,
        &Universe::simulateElectroweak,
        &Universe::simulateQuark,
        &Universe::simulateHadron,
        &Universe::simulateNucleosynthesis,
        &Universe::simulateRecombination,
        &Universe::simulateStarFormation,
        &Universe::simulateSolarSystem,
        &Universe::simulateEarth,
        &Universe::simulateLife,
        &Universe::simulateDNA,
        &Universe::simulatePresent,
    };

    for (auto fn : epochFns) {
        auto state = (this->*fn)();
        history_.push_back(state);
        if (onEpoch) onEpoch(state);
    }
}

EpochState Universe::runEpoch(int epochIndex) {
    using EpochFn = EpochState (Universe::*)();
    EpochFn epochFns[] = {
        &Universe::simulatePlanck,
        &Universe::simulateInflation,
        &Universe::simulateElectroweak,
        &Universe::simulateQuark,
        &Universe::simulateHadron,
        &Universe::simulateNucleosynthesis,
        &Universe::simulateRecombination,
        &Universe::simulateStarFormation,
        &Universe::simulateSolarSystem,
        &Universe::simulateEarth,
        &Universe::simulateLife,
        &Universe::simulateDNA,
        &Universe::simulatePresent,
    };

    if (epochIndex < 0 || epochIndex >= 13) {
        return snapshot("Unknown", "Invalid epoch index");
    }

    return (this->*epochFns[epochIndex])();
}

// ============================================================
// Epoch implementations
// ============================================================

EpochState Universe::simulatePlanck() {
    tick_ = PLANCK_EPOCH;
    environment.update(tick_);
    quantumField.temperature = environment.currentTemperature();

    // At the Planck epoch, all forces are unified.
    // Vacuum fluctuations create the initial particle content.
    for (int i = 0; i < 20; ++i) {
        quantumField.vacuumFluctuation();
    }

    std::ostringstream details;
    details << "All four forces unified. "
            << "Vacuum fluctuations create " << quantumField.particles.size()
            << " virtual particles. "
            << "Vacuum energy = " << std::scientific << std::setprecision(2)
            << quantumField.vacuumEnergy;

    return snapshot("Planck", details.str());
}

EpochState Universe::simulateInflation() {
    tick_ = INFLATION_EPOCH;
    environment.update(tick_);
    quantumField.temperature = environment.currentTemperature();

    // Inflation: exponential expansion, quantum fluctuations seed structure.
    // Massive pair production from inflaton field decay.
    for (int i = 0; i < 50; ++i) {
        double energy = quantumField.temperature * K_B * 10.0;
        quantumField.pairProduction(energy);
        quantumField.vacuumFluctuation();
    }

    // Evolve the field
    quantumField.evolve(1.0);

    std::ostringstream details;
    details << "Exponential expansion by factor ~e^60. "
            << "Quantum fluctuations seed large-scale structure. "
            << quantumField.totalCreated << " particles created, "
            << quantumField.particles.size() << " surviving.";

    return snapshot("Inflation", details.str());
}

EpochState Universe::simulateElectroweak() {
    tick_ = ELECTROWEAK_EPOCH;
    environment.update(tick_);
    quantumField.temperature = environment.currentTemperature();

    // Electroweak symmetry breaking: W/Z bosons become massive
    // More pair production at high energy
    for (int i = 0; i < 30; ++i) {
        double energy = quantumField.temperature * K_B * 5.0;
        quantumField.pairProduction(energy);
    }

    // Some annihilation (matter-antimatter)
    int annihilations = 0;
    for (size_t i = 0; i + 1 < quantumField.particles.size() && annihilations < 10; ++i) {
        if (quantumField.particles[i].type == ParticleType::Electron) {
            for (size_t j = i + 1; j < quantumField.particles.size(); ++j) {
                if (quantumField.particles[j].type == ParticleType::Positron) {
                    quantumField.annihilate(i, j);
                    ++annihilations;
                    break;
                }
            }
        }
    }

    quantumField.evolve(1.0);

    std::ostringstream details;
    details << "Higgs field activates, W/Z bosons gain mass. "
            << "EM and weak forces separate. "
            << annihilations << " matter-antimatter annihilations. "
            << quantumField.particles.size() << " particles remain.";

    return snapshot("Electroweak", details.str());
}

EpochState Universe::simulateQuark() {
    tick_ = QUARK_EPOCH;
    environment.update(tick_);
    quantumField.temperature = environment.currentTemperature();

    // Quark-gluon plasma era
    // Generate quark pairs
    for (int i = 0; i < 40; ++i) {
        double energy = M_PROTON * C * C * 3.0;
        quantumField.pairProduction(energy);
    }

    quantumField.evolve(1.0);

    // Count quarks
    int ups = 0, downs = 0;
    for (auto& p : quantumField.particles) {
        if (p.type == ParticleType::Up) ++ups;
        if (p.type == ParticleType::Down) ++downs;
    }

    std::ostringstream details;
    details << "Quark-gluon plasma: free quarks and gluons. "
            << ups << " up quarks, " << downs << " down quarks, "
            << quantumField.particles.size() << " total particles.";

    return snapshot("Quark", details.str());
}

EpochState Universe::simulateHadron() {
    tick_ = HADRON_EPOCH;
    environment.update(tick_);
    quantumField.temperature = T_QUARK_HADRON * 0.5; // below QCD transition

    // Quark confinement: quarks bind into protons and neutrons
    auto hadrons = quantumField.quarkConfinement();

    quantumField.evolve(1.0);

    int protons = 0, neutrons = 0;
    for (auto& p : quantumField.particles) {
        if (p.type == ParticleType::Proton)  ++protons;
        if (p.type == ParticleType::Neutron) ++neutrons;
    }

    std::ostringstream details;
    details << "Quarks confined into hadrons. "
            << hadrons.size() << " new hadrons formed: "
            << protons << " protons, " << neutrons << " neutrons. "
            << quantumField.particles.size() << " total particles.";

    return snapshot("Hadron", details.str());
}

EpochState Universe::simulateNucleosynthesis() {
    tick_ = NUCLEOSYNTHESIS_EPOCH;
    environment.update(tick_);
    quantumField.temperature = T_NUCLEOSYNTHESIS;
    atomicSystem.temperature = T_NUCLEOSYNTHESIS;

    // Count protons and neutrons in the quantum field
    int protons = 0, neutrons = 0;
    for (auto& p : quantumField.particles) {
        if (p.type == ParticleType::Proton)  ++protons;
        if (p.type == ParticleType::Neutron) ++neutrons;
    }

    // Big Bang nucleosynthesis: form H and He
    auto nuclei = atomicSystem.nucleosynthesis(protons, neutrons);

    int hydrogen = 0, helium = 0;
    for (auto& a : atomicSystem.atoms) {
        if (a.atomicNumber == 1) ++hydrogen;
        if (a.atomicNumber == 2) ++helium;
    }

    std::ostringstream details;
    details << "Primordial nucleosynthesis: "
            << hydrogen << " hydrogen, " << helium << " helium nuclei. "
            << "~75% H, ~25% He by mass (cosmological ratio). "
            << nuclei.size() << " nuclei formed.";

    return snapshot("Nucleosynthesis", details.str());
}

EpochState Universe::simulateRecombination() {
    tick_ = RECOMBINATION_EPOCH;
    environment.update(tick_);
    quantumField.temperature = T_RECOMBINATION;
    atomicSystem.temperature = T_RECOMBINATION;

    // Add electrons to the quantum field for recombination
    int protonCount = 0;
    for (auto& p : quantumField.particles) {
        if (p.type == ParticleType::Proton) ++protonCount;
    }

    // Ensure we have electrons for recombination
    int electronCount = 0;
    for (auto& p : quantumField.particles) {
        if (p.type == ParticleType::Electron) ++electronCount;
    }

    // Add more electrons if needed
    while (electronCount < protonCount) {
        Particle electron;
        electron.type = ParticleType::Electron;
        electron.spin = Spin::Up;
        quantumField.particles.push_back(electron);
        ++electronCount;
    }

    // Electrons captured by nuclei -> neutral atoms
    auto newAtoms = atomicSystem.recombination(quantumField);

    std::ostringstream details;
    details << "Electrons captured by nuclei: universe becomes transparent. "
            << newAtoms.size() << " new neutral atoms formed. "
            << atomicSystem.atoms.size() << " total atoms. "
            << "Cosmic Microwave Background released (T=" << T_CMB << "K today).";

    return snapshot("Recombination", details.str());
}

EpochState Universe::simulateStarFormation() {
    tick_ = STAR_FORMATION_EPOCH;
    environment.update(tick_);

    // Stellar nucleosynthesis: forge heavier elements
    auto heavyElements = atomicSystem.stellarNucleosynthesis(T_STELLAR_CORE);

    // Count elements
    int carbon = 0, oxygen = 0, nitrogen = 0, iron = 0;
    for (auto& a : atomicSystem.atoms) {
        if (a.atomicNumber == 6) ++carbon;
        if (a.atomicNumber == 7) ++nitrogen;
        if (a.atomicNumber == 8) ++oxygen;
        if (a.atomicNumber == 26) ++iron;
    }

    std::ostringstream details;
    details << "First stars ignite from gravitational collapse. "
            << "Stellar fusion forges heavy elements: "
            << carbon << " C, " << nitrogen << " N, "
            << oxygen << " O atoms. "
            << heavyElements.size() << " new heavy elements created. "
            << "Supernovae distribute elements across space.";

    return snapshot("Star Formation", details.str());
}

EpochState Universe::simulateSolarSystem() {
    tick_ = SOLAR_SYSTEM_EPOCH;
    environment.update(tick_);

    // More stellar nucleosynthesis to produce enough elements for chemistry
    atomicSystem.stellarNucleosynthesis(T_STELLAR_CORE);
    atomicSystem.stellarNucleosynthesis(T_STELLAR_CORE);

    // Ensure we have enough hydrogen for later chemistry
    // (stellar debris has lots of H)
    int hCount = 0;
    for (auto& a : atomicSystem.atoms) {
        if (a.atomicNumber == 1) ++hCount;
    }
    // Add more hydrogen from the primordial reservoir
    for (int i = 0; i < 50 && hCount < 60; ++i) {
        Atom h;
        h.init(1, 1);
        atomicSystem.atoms.push_back(h);
        ++hCount;
    }

    std::ostringstream details;
    details << "Solar nebula coalesces from stellar debris. "
            << "Protoplanetary disk forms. "
            << atomicSystem.atoms.size() << " atoms in the system. "
            << "Elements available for rocky planet formation.";

    return snapshot("Solar System", details.str());
}

EpochState Universe::simulateEarth() {
    tick_ = EARTH_EPOCH;
    environment.update(tick_);

    // Form water and simple molecules
    auto waters = chemicalSystem.formWater();
    auto methanes = chemicalSystem.formMethane();
    auto ammonias = chemicalSystem.formAmmonia();

    std::ostringstream details;
    details << "Earth accretes and differentiates. "
            << "Oceans form: " << waters.size() << " water molecules. "
            << methanes.size() << " CH4, " << ammonias.size() << " NH3. "
            << "Total molecules: " << chemicalSystem.molecules.size() << ". "
            << environment.summary();

    return snapshot("Earth", details.str());
}

EpochState Universe::simulateLife() {
    tick_ = LIFE_EPOCH;
    environment.update(tick_);

    // Prebiotic chemistry: form amino acids and nucleotides
    int aaFormed = 0, nucFormed = 0;
    for (int i = 0; i < 20; ++i) {
        int formed = chemicalSystem.catalyzedReaction(
            environment.currentTemperature(), true);
        aaFormed += chemicalSystem.aminoAcidCount;
        nucFormed += chemicalSystem.nucleotideCount;
        if (formed > 0) break; // got some biomolecules
    }

    // First self-replicating molecules emerge
    biosphere = std::make_unique<Biosphere>(3, 60);

    std::ostringstream details;
    details << "First self-replicating molecules appear! "
            << "Amino acids: " << chemicalSystem.aminoAcidCount
            << ", Nucleotides: " << chemicalSystem.nucleotideCount << ". "
            << biosphere->cells.size() << " protocells formed. "
            << "Avg fitness: " << std::fixed << std::setprecision(3)
            << biosphere->averageFitness() << ". "
            << environment.summary();

    return snapshot("Life", details.str());
}

EpochState Universe::simulateDNA() {
    tick_ = DNA_EPOCH;
    environment.update(tick_);

    if (!biosphere) {
        biosphere = std::make_unique<Biosphere>(5, 90);
    }

    // Run several generations of biological evolution
    double uvShielding = environment.atmosphere.uvShielding();
    double uvIntensity = environment.radiation.uvIntensity * (1.0 - uvShielding);
    double cosmicFlux  = environment.radiation.cosmicRayFlux;
    double bioEnergy   = environment.biologicalEnergy();

    for (int gen = 0; gen < 10; ++gen) {
        biosphere->step(bioEnergy, uvIntensity, cosmicFlux,
                        environment.currentTemperature());
    }

    std::ostringstream details;
    details << "DNA-based life dominates. Epigenetics emerge. "
            << biosphere->cells.size() << " cells, "
            << biosphere->generation << " generations. "
            << "Avg fitness: " << std::fixed << std::setprecision(3)
            << biosphere->averageFitness()
            << ", GC content: " << std::setprecision(3)
            << biosphere->averageGcContent() << ". "
            << "Mutations: " << biosphere->totalMutations()
            << ", Born: " << biosphere->totalBorn
            << ", Died: " << biosphere->totalDied << ". "
            << environment.summary();

    return snapshot("DNA Era", details.str());
}

EpochState Universe::simulatePresent() {
    tick_ = PRESENT_EPOCH;
    environment.update(tick_);

    if (!biosphere) {
        biosphere = std::make_unique<Biosphere>(10, 120);
    }

    // Run more generations to evolve complexity
    double uvShielding = environment.atmosphere.uvShielding();
    double uvIntensity = environment.radiation.uvIntensity * (1.0 - uvShielding);
    double cosmicFlux  = environment.radiation.cosmicRayFlux;
    double bioEnergy   = environment.biologicalEnergy();

    for (int gen = 0; gen < 20; ++gen) {
        biosphere->step(bioEnergy, uvIntensity, cosmicFlux,
                        environment.currentTemperature());
    }

    // Final census
    int totalProteins = 0;
    int totalGenes = 0;
    for (auto& cell : biosphere->cells) {
        totalProteins += static_cast<int>(cell.proteins.size());
        totalGenes += static_cast<int>(cell.dna.genes.size());
    }

    std::ostringstream details;
    details << "Complex life, intelligence, and consciousness emerge. "
            << biosphere->cells.size() << " cells across "
            << biosphere->generation << " generations. "
            << "Avg fitness: " << std::fixed << std::setprecision(3)
            << biosphere->averageFitness() << ". "
            << totalProteins << " proteins, " << totalGenes << " genes active. "
            << "Total mutations: " << biosphere->totalMutations()
            << ", Born: " << biosphere->totalBorn
            << ", Died: " << biosphere->totalDied << ". "
            << environment.summary();

    return snapshot("Present", details.str());
}

} // namespace sim
